internal/exec/exec.go

```diff
@@ -0,0 +1,477 @@
+			return nil, err
+		}
+	}
+
+	e := &Exec{
+		path:       path,
+		cmdline:    cmdLine,
+		execConfig: eopts,
+	}
+
+	if err := e.setupStdio(); err != nil {
+		return nil, err
+	}
+
+	return e, nil
+}
+
+// Start starts the process with the path and cmdline specified when the Exec object was created. This does not wait for exit or release any resources,
+// a call to Wait must be made afterwards.
+func (e *Exec) Start() error {
+	argv0 := e.path
+	if len(e.dir) != 0 {
+		// Windows CreateProcess looks for argv0 relative to the current
+		// directory, and, only once the new process is started, it does
+		// Chdir(attr.Dir). We are adjusting for that difference here by
+		// making argv0 absolute.
+		var err error
+		argv0, err = joinExeDirAndFName(e.dir, e.path)
+		if err != nil {
+			return err
+		}
+	}
+
+	argv0p, err := windows.UTF16PtrFromString(argv0)
+	if err != nil {
+		return err
+	}
+
+	argvp, err := windows.UTF16PtrFromString(e.cmdline)
+	if err != nil {
+		return err
+	}
+
+	var dirp *uint16
+	if len(e.dir) != 0 {
+		dirp, err = windows.UTF16PtrFromString(e.dir)
+		if err != nil {
+			return err
+		}
+	}
+
+	siEx := new(winapi.StartupInfoEx)
+	siEx.Flags = windows.STARTF_USESTDHANDLES
+	pi := new(windows.ProcessInformation)
+
+	// Need EXTENDED_STARTUPINFO_PRESENT as we're making use of the attribute list field.
+	flags := uint32(windows.CREATE_UNICODE_ENVIRONMENT) | windows.EXTENDED_STARTUPINFO_PRESENT | e.execConfig.processFlags
+
+	// Allocate an attribute list that's large enough to do the operations we care about
+	// 1. Assigning to a job object at creation time
+	// 2. Pseudo console setup if one was requested.
+	// 3. Inherit only stdio handles if ones were requested.
+	e.attrList, err = winapi.NewProcThreadAttributeList(3)
+	if err != nil {
+		return fmt.Errorf("failed to initialize process thread attribute list: %w", err)
+	}
+	siEx.ProcThreadAttributeList = e.attrList
+
+	// Need to know whether the process needs to inherit stdio handles.
+	inheritHandles := e.stdioProcEnd[0] != nil || e.stdioProcEnd[1] != nil || e.stdioProcEnd[2] != nil
+	if inheritHandles {
+		var handles []uintptr
+		for _, file := range e.stdioProcEnd {
+			if file.Fd() != uintptr(syscall.InvalidHandle) {
+				handles = append(handles, file.Fd())
+			}
+		}
+
+		// Set up the process to only inherit stdio handles and nothing else.
+		err = winapi.UpdateProcThreadAttribute(
+			siEx.ProcThreadAttributeList,
+			0,
+			windows.PROC_THREAD_ATTRIBUTE_HANDLE_LIST,
+			unsafe.Pointer(&handles[0]),
+			uintptr(len(handles))*unsafe.Sizeof(handles[0]),
+			nil,
+			nil,
+		)
+		if err != nil {
+			return err
+		}
+
+		// Assign the handles to the startupinfos stdio fields.
+		if e.stdioProcEnd[0] != nil {
+			siEx.StdInput = windows.Handle(e.stdioProcEnd[0].Fd())
+		}
+		if e.stdioProcEnd[1] != nil {
+			siEx.StdOutput = windows.Handle(e.stdioProcEnd[1].Fd())
+		}
+		if e.stdioProcEnd[2] != nil {
+			siEx.StdErr = windows.Handle(e.stdioProcEnd[2].Fd())
+		}
```

Comment:
Why do we need to set these handles both in attribute list and in startupInfoEx struct?

Reply:
So the AttributeList portion tells CreateProcess what handles we would like to inherit, which is only the stdio handles (all I could think of at least). Setting them here is what tells CreateProcess that these are the handles that we'd like to use for stdout/err/in. If we didn't do the attributelist setup I believe any handles that were marked as inheritable would be inherited so long as we have the [bInheritHandles](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessw#syntax) parameter set to true. We're just making sure we're only inheriting what we need to and nothing more.

Reply:
Added a comment about this